File: projectNew/old_src/TTA/InquisitionListener.java
package com.f14.TTA.listener.event;

import com.f14.TTA.TTAGameMode;
import com.f14.TTA.TTAPlayer;
import com.f14.TTA.component.ability.EventAbility;
import com.f14.TTA.component.card.TTACard;
import com.f14.TTA.component.card.TechCard;
import com.f14.TTA.component.param.PopParam;
import com.f14.TTA.consts.CardSubType;
import com.f14.TTA.consts.CardType;
import com.f14.TTA.consts.TTACmdString;
import com.f14.TTA.consts.TTAGameCmd;
import com.f14.TTA.listener.TTAEventListener;
import com.f14.bg.action.BgAction;
import com.f14.bg.exception.BoardGameException;
import com.f14.bg.listener.InterruptParam;
import com.f14.bg.utils.CheckUtils;
import com.f14.utils.CommonUtil;

import java.util.HashMap;
import java.util.Map;

public class InquisitionListener extends TTAEventListener {

    public InquisitionListener(EventAbility ability, TTAPlayer trigPlayer) {
        super(ability, trigPlayer);
    }

    @Override
    protected boolean beforeListeningCheck(TTAGameMode gameMode, TTAPlayer p) {
        return !this.canPass(p);
    }

    @Override
    protected void beforeStartListen(TTAGameMode gameMode) throws BoardGameException {
        super.beforeStartListen(gameMode);
        // 为所有玩家创建人口参数
        gameMode.inquisitionPosition = this.trigPlayer.getPosition();
        for (TTAPlayer player : gameMode.getRealPlayers()) {
            int limit = player.goverment.buildingLimit;
            InquisitionParam param = new InquisitionParam();
            for (CardSubType subType : new CardSubType[]{CardSubType.LAB, CardSubType.LIBRARY}) {
                int shouldDestroy = Math.max(0, player.getBuildingNumber(subType) - limit + 1);
                param.shouldDestroys.put(subType, shouldDestroy);
                for (TTACard c : player.getBuildingsBySubType(subType)) {
                    if (c.getAvailableCount() == limit) {
                        try {
                            param.destory((TechCard) c);
                            gameMode.getGame().refreshDecreasePopulation(player, (TechCard) c, 1);
                        } catch (Exception e) {
                            e.printStackTrace();
                        }
                    }
                }
            }
            this.setParam(player.getPosition(), param);
        }
    }

    protected boolean canPass(TTAPlayer player) {
        InquisitionParam param = this.getParam(player);
        for (Integer n : param.shouldDestroys.values()) {
            if (n != 0) {
                return false;
            }
        }
        return true;
    }


    @Override
    public InterruptParam createInterruptParam() {
        InterruptParam res = new InterruptParam();
        for (TTAPlayer p : this.listeningPlayers) {
            InquisitionParam param = this.getParam(p);
            res.set(p.getPosition(), param.param);
        }
        return res;
    }

    @Override
    protected void doAction(TTAGameMode gameMode, BgAction action) throws BoardGameException {
        TTAPlayer player = action.getPlayer();
        InquisitionParam param = this.getParam(player);
        boolean confirm = action.getAsBoolean("confirm");
        if (confirm) {
            String cardId = action.getAsString("cardId");
            TTACard card = player.getBuildings().getCard(cardId);
            CheckUtils.check(card.cardType != CardType.BUILDING, "不能选择这张牌!");
            CheckUtils.check(card.cardSubType != CardSubType.LIBRARY && card.cardSubType != CardSubType.LAB,
                    "不能选择这张牌!");
            param.destory((TechCard) card);
            gameMode.getGame().refreshDecreasePopulation(player, (TechCard) card, 1);
            if (this.canPass(player)) {
                this.setPlayerResponsed(gameMode, player.getPosition());
            }
        } else {
            // 判断玩家是否可以结束
            CheckUtils.check(!this.canPass(player), this.getMsg(player));
            this.setPlayerResponsed(gameMode, player.getPosition());
        }
    }


    @Override
    protected String getActionString() {
        return TTACmdString.ACTION_DESTORY;
    }

    @Override
    protected String getMsg(TTAPlayer player) {
        InquisitionParam param = this.getParam(player);
        String msg = "你需要摧毁{0}个图书馆和{1}个实验室,请选择!";
        msg = CommonUtil.getMsg(msg, param.shouldDestroys.get(CardSubType.LIBRARY),
                param.shouldDestroys.get(CardSubType.LAB));
        return msg;
    }

    @Override
    protected int getValidCode() {
        return TTAGameCmd.GAME_CODE_EVENT_DESTORY;
    }

    class InquisitionParam {

        public Map<CardSubType, Integer> shouldDestroys = new HashMap<>();

        public PopParam param = new PopParam();

        /**
         * 拆除建筑
         *
         * @param card
         * @throws BoardGameException
         */
        void destory(TechCard card) throws BoardGameException {
            CheckUtils.check(card.getWorkers() == 0, "这张牌上没有工人!");
            Integer shouldDestroy = shouldDestroys.get(card.cardSubType);
            if (shouldDestroy != null && shouldDestroy > 0) {
                shouldDestroy -= 1;
                shouldDestroys.put(card.cardSubType, shouldDestroy);
                param.destory(card);
                return;
            }
            throw new BoardGameException("不能选择这张牌!");
        }

    }

}

File: projectNew/src/com/f14/TTA/listener/event/InquisitionListener.kt
package com.f14.TTA.listener.event

import com.f14.TTA.TTAGameMode
import com.f14.TTA.TTAPlayer
import com.f14.TTA.component.ability.EventAbility
import com.f14.TTA.component.card.TechCard
import com.f14.TTA.component.param.PopParam
import com.f14.TTA.consts.CardSubType
import com.f14.TTA.consts.CardType
import com.f14.TTA.consts.TTACmdString
import com.f14.TTA.consts.TTAGameCmd
import com.f14.TTA.listener.TTAEventListener
import com.f14.bg.action.BgAction.GameAction
import com.f14.bg.exception.BoardGameException
import com.f14.bg.listener.InterruptParam
import com.f14.bg.utils.CheckUtils
import com.f14.utils.CommonUtil
import java.util.*
import kotlin.math.max

/**
 * Created by 吹风奈奈 on 2017/7/20.
 */

class InquisitionListener(gameMode: TTAGameMode, ability: EventAbility, trigPlayer: TTAPlayer) : TTAEventListener(gameMode, ability, trigPlayer) {

    override fun beforeListeningCheck(player: TTAPlayer): Boolean {
        return !this.canPass(player)
    }

    @Throws(BoardGameException::class)
    override fun beforeStartListen() {
        super.beforeStartListen()
        // 为所有玩家创建人口参数
        for (player in gameMode.realPlayers) {
            val limit = player.government!!.buildingLimit
            val param = InquisitionParam()
            for (subType in arrayOf(CardSubType.LAB, CardSubType.LIBRARY)) {
                val shouldDestroy = max(0, player.getBuildingNumber(subType) - limit + 1)
                param.shouldDestroys[subType] = shouldDestroy
                player.getBuildingsBySubType(subType).filter { it.availableCount == limit }.map { param.destory(it as TechCard) }.forEach { gameMode.game.refreshDecreasePopulation(player, it, 1) }
            }
            this.setParam(player.position, param)
        }
    }

    private fun canPass(player: TTAPlayer): Boolean {
        val param = this.getParam<InquisitionParam>(player)
        return param.shouldDestroys.values.none { it != 0 }
    }


    override fun createInterruptParam(): InterruptParam {
        val res = super.createInterruptParam()
        for (p in this.listeningPlayers) {
            val param = this.getParam<InquisitionParam>(p)
            res[p.position] = param.param
        }
        return res
    }

    @Throws(BoardGameException::class)
    override fun doAction(action: GameAction) {
        val player = action.getPlayer<TTAPlayer>()
        val param = this.getParam<InquisitionParam>(player)
        val confirm = action.getAsBoolean("confirm")
        if (confirm) {
            val cardId = action.getAsString("cardId")
            val card = player.buildings.getCard(cardId)
            CheckUtils.check(card.cardType != CardType.BUILDING, "不能选择这张牌!")
            CheckUtils.check(card.cardSubType != CardSubType.LIBRARY && card.cardSubType != CardSubType.LAB, "不能选择这张牌!")
            param.destory(card as TechCard)
            gameMode.game.refreshDecreasePopulation(player, card, 1)
            if (this.canPass(player)) {
                this.setPlayerResponsed(player)
            }
        } else {
            // 判断玩家是否可以结束
            CheckUtils.check(!this.canPass(player), this.getMsg(player))
            this.setPlayerResponsed(player)
        }
    }


    override val actionString: String
        get() = TTACmdString.ACTION_DESTORY

    override fun getMsg(player: TTAPlayer): String {
        val param = this.getParam<InquisitionParam>(player)
        var msg = "你需要摧毁{0}个图书馆和{1}个实验室,请选择!"
        msg = CommonUtil.getMsg(msg, param.shouldDestroys[CardSubType.LIBRARY]
                ?: 0, param.shouldDestroys[CardSubType.LAB] ?: 0)
        return msg
    }

    override val validCode: Int
        get() = TTAGameCmd.GAME_CODE_EVENT_DESTORY

    internal inner class InquisitionParam {
        var shouldDestroys: MutableMap<CardSubType, Int> = HashMap()
        var param = PopParam()

        /**
         * 拆除建筑
         * @param card
         * @throws BoardGameException
         */
        @Throws(BoardGameException::class)
        fun destory(card: TechCard): TechCard {
            CheckUtils.check(card.workers == 0, "这张牌上没有工人!")
            var shouldDestroy: Int = shouldDestroys[card.cardSubType] ?: throw BoardGameException("不能选择这张牌!")
            if (shouldDestroy == 0) throw BoardGameException("不能选择这张牌!")
            shouldDestroy -= 1
            shouldDestroys[card.cardSubType] = shouldDestroy
            param.destory(card)
            return card
        }

    }

}
